### PR TITLE
Logic of Etherpad (Phase 1)

### PR DESCRIPTION
**Pull request by anantjain6**
Following thing are implemented:-

1. "Test Connection" button
Screenshot from 2019-06-14 01-55-40

2. hook_requirements() to check presence of Etherpad client library and jQuery Plugin, and give requirement error if not present
_site.com/admin/report/status_
Screenshot from 2019-06-13 22-24-16

3.1. Using Editor API of Drupal's core module "Text Editor" I make the "Etherpad" editor discoverable in Drupal system
_site.com/admin/config/content/formats_ Add or Configure any Text Format and see "Etherpad" as an available text editor
Screenshot from 2019-06-14 01-50-18

3.2. Added skeleton for Text Editor of Drupal
Screenshot from 2019-06-14 02-06-38

4. Added hook_library_info_build() to include Etherpad jQuery Plugin
_It is a challenging task
First I have 2 option to use Drupal's native library discovery or use Libraries module. Then iI chosen Native way
Second I need to understand which type of library I'm including to shortlist the best way to include it. It is a dynamic library. The library may be static or dynamic
Third, from all the shortlisted way I have chosen hook_library_info_build()_

### CODE REVIEW

**Comment by ankitjain28may** [comment]
@anantjain6 No preview link?
Create also a screencast so that i can install different libraries and test it out.

**Comment by anantjain6** [comment]
Screencast https://youtu.be/Ij9SRZBVxVQ till 04:40 for this PR

**Comment by anantjain6** [comment]
Preview: https://master-sh17dlrdxxhpywypabojiyyuku3zqz4x.tugboat.qa/
Username: admin
Password: pass

**Comment by ankitjain28may** [comment]
@anantjain6 Is this preview working, Did you configure nodejs and etherpad client libraries?

**Comment by anantjain6** [comment]
Preview is for this PR
which contains only phase 1 logic
Hence in it node-js is not needed

I have intentionally not added libraries in this preview to show requirement errors

**Comment by ankitjain28may** [comment]
@anantjain6 How will i review it?

**Comment by anantjain6** [comment]
Added node-js & Etherpad

Preview: https://dev6-i80fekg0edomlhwguvfdij42vuqfjxbd.tugboat.qa/
Username: admin
Password: pass

In /admin/config/collaborative-editors/etherpad
I have saved the Etherpad API URL & Key

**Comment by ankitjain28may** [comment]
and what about other etherpad libraries?

**Comment by anantjain6** [comment]
They are also installed

**Comment by ankitjain28may** [comment]
@anantjain6 In the preview, Etherpad is not getting load on the client side, may be you have missed Etherpad client side libraries. Do try it first on preview and let me know once its working there.

**Comment by anantjain6** [comment]
This PR contains half logic of Etherpad
The complete integration is done in next PR
https://github.com/dbjpanda/collaborative_editors/pull/8

Please go through the description of PR

File: modules/ce_etherpad/ce_etherpad.install [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    $requirements['etherpad_client'] = [
+      'title' => t('Etherpad client library'),
+    ];
+    if (class_exists('EtherpadLite\Client')) {
+      $requirements['etherpad_client']['value'] = t('Installed');
+      $requirements['etherpad_client']['severity'] = REQUIREMENT_OK;
+    }
+    else {
+      $requirements['etherpad_client']['value'] = t('Not installed');
+      $requirements['etherpad_client']['description'] = t('Either Http Client library is not installed. Run "composer require 0x46616c6b/etherpad-lite-client" in root directory of Drupal to install Etherpad Http Client libreary.');
```
Comment (ankitjain28may) [inline modules/ce_etherpad/ce_etherpad.install]: library typo

File: modules/ce_etherpad/ce_etherpad.install [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+      $requirements['etherpad_client']['severity'] = REQUIREMENT_ERROR;
+    }
+
+    // Check for the jquery plugin.
+    $requirements['etherpad_plugin'] = [
```
Comment (ankitjain28may) [inline modules/ce_etherpad/ce_etherpad.install:30]: The error is not showing if the etherpad_plugin is not installed

File: modules/ce_etherpad/ce_etherpad.install [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+      'title' => t('Etherpad jQuery plugin'),
+    ];
+    if (file_exists(DRUPAL_ROOT . '/libraries/ce_etherpad/js/etherpad.js')) {
+      $requirements['etherpad_plugin']['value'] = DRUPAL_ROOT . '/libraries/ce_etherpad/js/etherpad.js';
+      $requirements['etherpad_plugin']['severity'] = REQUIREMENT_OK;
+    }
+    else {
+      $requirements['etherpad_plugin']['value'] = t('Not installed');
+      $requirements['etherpad_plugin']['description'] = t('Download <a href="@url">Etherpad jQuery Plugin</a>, extract in DRUPAL_ROOT/libraries directory and rename the folder "etherpad-lite-jquery-plugin-master" to "ce_etherpad".', array('@url' => 'https://github.com/ether/etherpad-lite-jquery-plugin/archive/master.zip'));
```
Comment (ankitjain28may) [inline modules/ce_etherpad/ce_etherpad.install]: DRUPAL_ROOT should actually return the path in error message

File: modules/ce_etherpad/src/Form/EtherpadSettingsForm.php [hunk 4]
```diff
@@ -49,6 +50,11 @@ public function buildForm(array $form, FormStateInterface $form_state) {
       '#description' => $this->t('Enter Etherpad API Key. You can find API Key in APIKEY.txt on root directory of Etherpad.'),
       '#default_value' => $config->get('etherpad_api_key'),
     ];
+    $form['test_connection'] = array(
```
Comment (ankitjain28may) [inline modules/ce_etherpad/src/Form/EtherpadSettingsForm.php]: array should be defined as []

File: modules/ce_etherpad/src/Form/EtherpadSettingsForm.php [hunk 5]
```diff
@@ -49,6 +50,11 @@ public function buildForm(array $form, FormStateInterface $form_state) {
...
+      '#type' => 'submit',
+      '#value' => t('Test connection'),
```
Comment (ankitjain28may) [inline modules/ce_etherpad/src/Form/EtherpadSettingsForm.php]: $this->t

File: modules/ce_etherpad/src/Form/EtherpadSettingsForm.php [hunk 6]
```diff
@@ -49,6 +50,11 @@ public function buildForm(array $form, FormStateInterface $form_state) {
...
+      '#submit' => array('::testConnection'),
```
Comment (ankitjain28may) [inline modules/ce_etherpad/src/Form/EtherpadSettingsForm.php]: Test Connection and Submit button should be in the same row

Comment (ankitjain28may) [reply modules/ce_etherpad/src/Form/EtherpadSettingsForm.php]: When I make changes to API_URL and click on test connection button, it loads the default value in the API_URL, not the updated value.

**Review by ankitjain28may — CHANGES_REQUESTED** [review]

File: modules/ce_etherpad/ce_etherpad.install (hunk 2 above)
Comment (anantjain6) [reply modules/ce_etherpad/ce_etherpad.install:30]: I have re-check, this error is showing if etherpad jQuery plugin not exists

Screenshot from 2019-06-25 16-54-14

**Comment by anantjain6** [comment]
Suggestions implemented with commit 96a5cc4
Preview: https://dev6-i80fekg0edomlhwguvfdij42vuqfjxbd.tugboat.qa
Username: admin
Password: pass

File: modules/ce_etherpad/ce_etherpad.install (hunk 2 above)
Comment (ankitjain28may) [reply modules/ce_etherpad/ce_etherpad.install:30]: It should throw an error on the same page i.e configuration page like it is throwing for the composer package

Comment (anantjain6) [reply modules/ce_etherpad/ce_etherpad.install:30]: At "Test Connection"?
In testing connection, composer package is necessary
jQuery plugin is used to load Etherpad

The reason why I don't prefer this error to be there
1. jQuery plugin is part of our Requirement, And Drupal has dedicated a place for requirement error So, I'm throwing REQUIREMENT_ERROR which you can see in the screenshot
2. Repetition of the same check at two places.

Comment (ankitjain28may) [reply modules/ce_etherpad/ce_etherpad.install:30]: At least show this error during plugin installation like a warning else user check when its plugin is not working and then he will see the logs.

Comment (anantjain6) [reply modules/ce_etherpad/ce_etherpad.install:30]: Implemented with commit 4c16164

**Review by ankitjain28may — CHANGES_REQUESTED** [review]
What about the php unit tests?

**Comment by ankitjain28may** [comment]
@anantjain6 Why are you messing in this PR, make separate PR for that.

**Comment by anantjain6** [comment]
Okay

**Comment by ankitjain28may** [comment]
@anantjain6 It contains travis.yml file commits, I am closing this PR, raise a new PR with only the required code and just one commit, always squash your commits before raising a PR.

**Comment by ankitjain28may** [comment]
@anantjain6  Did you check this?

**Pull request closed without merging**